apps/src/codebridge/FilePreview/HTMLPreview.tsx

```diff
@@ -1,69 +1,133 @@
+    setLevelLoading(false);
+  });
 
-        const styleFile = Object.values(files).find(
-          f => f.name === styleName && f.folderId === folderId
+  useEffect(() => {
+    const handleMessage = (event: MessageEvent) => {
+      if (event.origin !== previewUrl) {
+        return;
+      }
+      if (event.data.type === IframeMessageType.IFRAME_READY) {
+        setIsIframeLoaded(true);
+        iframeRef.current?.contentWindow?.postMessage(
+          {type: IframeMessageType.CHANGE_FILE_URL_BAR, fileName: currentFile},
+          previewUrl
         );
-
-        return `
-          <style>
-            ${styleFile?.contents}
-          </style>
-      `;
+      } else if (
+        event.data.type === IframeMessageType.FILE_UPDATED &&
+        event.origin === previewUrl
+      ) {
+        setCurrentFile(event.data.fileName);
       }
-    );
+    };
+
+    window.addEventListener('message', handleMessage);
+    return () => window.removeEventListener('message', handleMessage);
+  }, [previewUrl, currentFile]);
 
-    return contents;
-  }, [files, folders, file]);
+  useEffect(() => {
+    const debouncedUpdate = setTimeout(() => {
+      iframeRef.current?.contentWindow?.postMessage(
+        {type: IframeMessageType.CHANGE_FILE_URL_BAR, fileName: currentFile},
+        previewUrl
+      );
+    }, 300);
+
+    return () => clearTimeout(debouncedUpdate);
+  }, [currentFile, previewUrl]);
+
+  useEffect(() => {
+    if (levelLoading) {
+      // If the level is currently loading, we skip sending a potentially outdated source.
+      return;
+    }
+    if (sourceLevelId.current !== levelProperties.id) {
+      // If we have a new level id, update the source immediately.
+      setDebouncedSource(source);
+      sourceLevelId.current = levelProperties.id;
+    } else {
+      // Set a timeout to send the debounced value after 500ms
+      const debouncedSourceSetter = setTimeout(() => {
+        setDebouncedSource(source);
+      }, 500);
+
+      // Cleanup the timeout if source or level changes before 500ms has elapsed.
+      return () => {
+        clearTimeout(debouncedSourceSetter);
+      };
+    }
+  }, [source, levelProperties.id, levelLoading]);
+
+  useEffect(() => {
+    if (isIframeLoaded && iframeRef.current && debouncedSource && previewUrl) {
+      iframeRef.current.contentWindow?.postMessage(
+        {
+          type: IframeMessageType.SET_SOURCE,
+          source: debouncedSource,
+        },
+        previewUrl
+      );
+    }
+  }, [previewUrl, debouncedSource, isIframeLoaded]);
 
   return (
     <div className={moduleStyles.previewContainer}>
-      {file && (
-        <iframe
-          sandbox=""
-          allow="self"
-          title="Web Preview"
-          ref={iframeRef}
-          id="preview"
-          style={{width: '100%', height: '100%', backgroundColor: 'white'}}
-          srcDoc={srcdoc}
-          className={moduleStyles.iframePreview}
+      <div>
+        <TextField
```

Comment:
Just curious - will this eventually be a dropdown?

Reply:
No the idea is for it to act like a "url bar" that students can type into. We can discuss with design/curriculum what should happen if they type in a non-html file. If you try to view a non-html file in a real browser it will just show you the text of the file, which seems reasonable for now.